File: backend/app/application/services/smartread/csv_transformer.py
"""SmartRead CSV変換器.

横持ちCSV→縦持ちCSV変換とバリデーションを行う。
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any


logger = logging.getLogger(__name__)


@dataclass
class ValidationError:
    """バリデーションエラー."""

    row: int
    field: str
    message: str
    value: str | None = None


@dataclass
class TransformResult:
    """変換結果."""

    long_data: list[dict[str, Any]]
    errors: list[ValidationError] = field(default_factory=list)


# 共通項目（全ての明細行にコピーされる）
COMMON_FIELDS = [
    "ファイル名",
    "ページ番号",
    "テンプレート名",
    "発行日",
    "納品書No",
    "発注者",
    "発注事業所",
    "出荷場所名称",
    "納入日",
    "便",
]

# 明細項目（番号付き、例: 材質コード1, 材質コード2）
DETAIL_FIELDS = [
    "材質コード",
    "材質サイズ",
    "単位",
    "納入量",
    "納入量小数点",
    "アイテム",
    "購買",
    "次区",
]

# サブ明細項目（例: Lot No1-1, Lot No1-2）
SUB_DETAIL_FIELDS = [
    "Lot No",
    "梱包数",
    "数量",
]


class SmartReadCsvTransformer:
    """SmartRead CSV変換器."""

    def __init__(self, max_details: int = 20) -> None:
        """初期化.

        Args:
            max_details: 検出する明細の最大数
        """
        self.max_details = max_details

    def transform_to_long(
        self,
        wide_data: list[dict[str, Any]],
        skip_empty: bool = True,
    ) -> TransformResult:
        """横持ち→縦持ち変換.

        1行に複数明細が横に並ぶデータを、1行1明細の縦持ちに変換する。

        Args:
            wide_data: 横持ちデータ（1行=1ページ）
            skip_empty: 空明細をスキップするか

        Returns:
            変換結果
        """
        logger.info(f"[Transformer] Starting transformation of {len(wide_data)} wide rows")
        if wide_data:
            logger.info(
                f"[Transformer] First row keys ({len(wide_data[0])} total): {list(wide_data[0].keys())[:15]}..."
            )

        long_data: list[dict[str, Any]] = []
        errors: list[ValidationError] = []

        for row_idx, row in enumerate(wide_data):
            # 共通項目を抽出
            common = self._extract_common_fields(row)
            logger.debug(
                f"[Transformer] Row {row_idx}: Common fields extracted: {list(common.keys())}"
            )

            # 共通項目のバリデーション
            common, row_errors = self._validate_common_fields(common, row_idx)
            errors.extend(row_errors)

            # 明細を抽出
            details = self._extract_details(row)
            logger.info(f"[Transformer] Row {row_idx}: {len(details)} details extracted")

            for detail_idx, detail in enumerate(details):
                # 空明細スキップ
                if skip_empty and self._is_empty_detail(detail):
                    logger.debug(
                        f"[Transformer] Row {row_idx}, Detail {detail_idx}: SKIPPED (empty)"
                    )
                    continue

                # 明細のバリデーション
                detail, detail_errors = self._validate_detail(detail, row_idx, detail_idx + 1)
                errors.extend(detail_errors)

                # 共通項目と明細をマージ
                long_row = {
                    **common,
                    "明細番号": detail_idx + 1,
                    **detail,
                }
                long_data.append(long_row)
                logger.debug(
                    f"[Transformer] Row {row_idx}, Detail {detail_idx}: ADDED to long_data"
                )

        logger.info(
            f"[Transformer] COMPLETE: {len(wide_data)} wide rows -> {len(long_data)} long rows, {len(errors)} errors"
        )
        return TransformResult(long_data=long_data, errors=errors)

    def _extract_common_fields(self, row: dict[str, Any]) -> dict[str, Any]:
        """共通項目を抽出."""
        common = {}
        normalized_row = self._create_normalized_row(row)
        for field_name in COMMON_FIELDS:
            if field_name in normalized_row:
                common[field_name] = self._normalize_value(normalized_row[field_name])
        return common

    def _extract_details(self, row: dict[str, Any]) -> list[dict[str, Any]]:
        """明細項目を抽出.

        横持ちの「材質コード1」「材質コード2」... を
        [{"材質コード": "...", ...}, {"材質コード": "...", ...}] に変換。
        """
        details: list[dict[str, Any]] = []

        normalized_row = self._create_normalized_row(row)

        for n in range(1, self.max_details + 1):
            detail: dict[str, Any] = {}

            # 通常の明細項目（材質コード1, 材質コード 1, 納入量1, etc）
            for field_name in DETAIL_FIELDS:
                # 複数のパターンを試行
                keys_to_try = [f"{field_name}{n}", f"{field_name} {n}"]
                if n == 1:
                    keys_to_try.append(field_name)

                for key in keys_to_try:
                    if key in normalized_row:
                        detail[field_name] = self._normalize_value(normalized_row[key])
                        break

            # 納入量と納入量小数点を結合
            self._combine_quantity_and_decimal(detail)

            # サブ明細項目（Lot No1-1, Lot No 1-1, Lot No-1, etc）
            for sub_field in SUB_DETAIL_FIELDS:
                for sub_n in range(1, 5):  # 最大4つのサブ明細
                    keys_to_try = [f"{sub_field}{n}-{sub_n}", f"{sub_field} {n}-{sub_n}"]
                    if n == 1:
                        keys_to_try.append(f"{sub_field}-{sub_n}")

                    for key in keys_to_try:
                        if key in normalized_row:
                            detail[f"{sub_field}{sub_n}"] = self._normalize_value(
                                normalized_row[key]
                            )
                            break

            # 明細が存在する場合のみ追加
            is_empty = self._is_empty_detail(detail)
            if not is_empty:
                details.append(detail)
                logger.debug(
                    f"[Transformer] Detail n={n}: NOT empty, extracted: {list(detail.keys())}"
                )
                # 番号付きが見つからず、かつn=1で番号なしでヒットした場合は縦持ちと判断して打ち切り
                is_vertical = any(
                    field_name in normalized_row
                    and f"{field_name}1" not in normalized_row
                    and f"{field_name} 1" not in normalized_row
                    for field_name in DETAIL_FIELDS
                )
                if is_vertical and n == 1:
                    logger.debug("[Transformer] Detected vertical format, stopping at n=1")
                    break
            else:
                logger.debug(f"[Transformer] Detail n={n}: EMPTY, detail contents: {detail}")

        logger.debug(f"[Transformer] Total details extracted: {len(details)}")
        return details

    def _combine_quantity_and_decimal(self, detail: dict[str, Any]) -> None:
        """納入量と納入量小数点を結合.

        納入量小数点の値（3桁想定）を納入量の小数部として結合する。
        結合後は納入量小数点フィールドは削除される。

        Args:
            detail: 明細データ（in-place修正）
        """
        base_quantity = detail.get("納入量", "").strip()
        decimal_part = detail.get("納入量小数点", "").strip()

        # 納入量小数点フィールドを削除（縦持ちには含めない）
        if "納入量小数点" in detail:
            del detail["納入量小数点"]

        # 小数点がある場合は結合
        if decimal_part:
            if base_quantity:
                detail["納入量"] = f"{base_quantity}.{decimal_part}"
            else:
                detail["納入量"] = f"0.{decimal_part}"

    def _is_empty_detail(self, detail: dict[str, Any]) -> bool:
        """空明細かどうかを判定.

        全ての明細項目が空または未設定の場合は空明細。
        """
        if not detail:
            return True

        for key, value in detail.items():
            if key.startswith("エラー_"):
                continue
            if value is None:
                continue
            if str(value).strip() != "":
                return False

        return True

    def _create_normalized_row(self, row: dict[str, Any]) -> dict[str, Any]:
        """キーを正規化したrowを返す."""
        return {self._normalize_key(key): value for key, value in row.items()}

    def _normalize_key(self, key: str) -> str:
        """列名を正規化.

        - 前後空白トリム
        - 全角数字→半角数字
        """
        trimmed = key.strip()
        return trimmed.translate(str.maketrans("０１２３４５６７８９", "0123456789"))

    def _normalize_value(self, value: Any) -> str:
        """値を正規化.

        - 全角→半角変換
        - 前後空白トリム
        - None → 空文字
        """
        if value is None:
            return ""

        s = str(value).strip()

        # 全角→半角変換（数字、英字、一部記号）
        s = s.translate(
            str.maketrans(
                "０１２３４５６７８９ＡＢＣＤＥＦＧＨＩＪＫＬＭＮＯＰＱＲＳＴＵＶＷＸＹＺａｂｃｄｅｆｇｈｉｊｋｌｍｎｏｐｑｒｓｔｕｖｗｘｙｚ",
                "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz",
            )
        )

        return s

    def _validate_common_fields(
        self, common: dict[str, Any], row_idx: int
    ) -> tuple[dict[str, Any], list[ValidationError]]:
        """共通項目のバリデーション."""
        errors: list[ValidationError] = []

        # 発行日のバリデーション
        if "発行日" in common:
            parsed, err = self._parse_date(common["発行日"])
            if err:
                errors.append(
                    ValidationError(
                        row=row_idx,
                        field="発行日",
                        message="日付形式が不正です",
                        value=common["発行日"],
                    )
                )
                common["エラー_発行日形式"] = 1
                common["発行日"] = ""  # 値をクリア
            else:
                common["発行日"] = parsed
                common["エラー_発行日形式"] = 0

        # 納入日のバリデーション
        if "納入日" in common:
            parsed, err = self._parse_date(common["納入日"])
            if err:
                errors.append(
                    ValidationError(
                        row=row_idx,
                        field="納入日",
                        message="日付形式が不正です",
                        value=common["納入日"],
                    )
                )
                common["エラー_納入日形式"] = 1
                common["納入日"] = ""  # 値をクリア
            else:
                common["納入日"] = parsed
                common["エラー_納入日形式"] = 0

        return common, errors

    def _validate_detail(
        self, detail: dict[str, Any], row_idx: int, detail_idx: int
    ) -> tuple[dict[str, Any], list[ValidationError]]:
        """明細項目のバリデーション."""
        errors: list[ValidationError] = []

        # 次区のバリデーション（英数字のみ）
        if "次区" in detail:
            jiku = detail["次区"]
            if jiku and not re.match(r"^[A-Za-z0-9]+$", jiku):
                errors.append(
                    ValidationError(
                        row=row_idx,
                        field=f"次区(明細{detail_idx})",
                        message="次区は英数字である必要があります",
                        value=jiku,
                    )
                )
                detail["エラー_次区形式"] = 1
            else:
                detail["エラー_次区形式"] = 0

        # 納入量のバリデーション（数値、非負）
        if "納入量" in detail:
            quantity = detail["納入量"]
            parsed, err = self._parse_quantity(quantity)
            if err:
                errors.append(
                    ValidationError(
                        row=row_idx,
                        field=f"納入量(明細{detail_idx})",
                        message="納入量は0以上の数値である必要があります",
                        value=quantity,
                    )
                )
                detail["エラー_納入量"] = 1
                # detail["納入量"] = parsed  # _parse_quantityはエラー時も値を返すのでそのまま
            else:
                detail["納入量"] = parsed
                detail["エラー_納入量"] = 0

        return detail, errors

    def _parse_date(self, value: str) -> tuple[str, bool]:
        """日付をパース.

        Args:
            value: 日付文字列

        Returns:
            (正規化された日付 YYYY/MM/DD, エラーフラグ)
        """
        if not value:
            return "", False

        # 正規化
        s = self._normalize_value(value)

        # 各種形式を試行
        formats = [
            "%Y/%m/%d",
            "%Y-%m-%d",
            "%Y年%m月%d日",
            "%Y.%m.%d",
            "%Y/%m",
            "%Y-%m",
            "%Y年%m月",
            "%Y年",
            "%Y",
        ]

        for fmt in formats:
            try:
                dt = datetime.strptime(s, fmt)
                # 範囲チェック
                if dt.year < 1900 or dt.year > 9999:
                    return "", True
                return dt.strftime("%Y/%m/%d"), False
            except ValueError:
                continue

        return "", True

    def _parse_quantity(self, value: str) -> tuple[str, bool]:
        """数量をパース.

        Args:
            value: 数量文字列

        Returns:
            (正規化された数量, エラーフラグ)
        """
        if not value:
            return "", False

        s = self._normalize_value(value)

        # カンマを削除（1,000 → 1000）
        s = s.replace(",", "")

        try:
            num = float(s)
            if num < 0:
                return s, True
            # 整数なら整数として返す
            if num == int(num):
                return str(int(num)), False
            return str(num), False
        except ValueError:
            return s, True

File: frontend/src/features/rpa/smartread/utils/csv-transformer.ts
/**
 * SmartRead CSV Transformer
 *
 * TypeScript port of the Python SmartReadCsvTransformer
 * Transforms wide (horizontal) CSV data to long (vertical) format
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import { logger } from "./logger";

export interface ValidationError {
  row: number;
  field: string;
  message: string;
  value: string | null;
}

export interface TransformResult {
  long_data: Array<Record<string, any>>;
  errors: ValidationError[];
}

// Common fields (copied to all detail rows)
const COMMON_FIELDS = [
  "ファイル名",
  "ページ番号",
  "テンプレート名",
  "発行日",
  "納品書No",
  "発注者",
  "発注事業所",
  "受注者",
  "出荷場所名称",
  "納入日",
  "便",
];

// Detail fields (numbered, e.g. 材質コード1, 材質コード2)
const DETAIL_FIELDS = ["材質コード", "材質サイズ", "単位", "納入量", "アイテム", "購買", "次区"];

// Sub-detail fields (e.g. Lot No1-1, Lot No1-2)
const SUB_DETAIL_FIELDS = ["Lot No", "梱包数"];

export class SmartReadCsvTransformer {
  private maxDetails: number;

  constructor(maxDetails: number = 20) {
    this.maxDetails = maxDetails;
  }

  /**
   * Transform wide format to long format
   *
   * Converts data where multiple details are in one row (horizontal)
   * to data where each detail is a separate row (vertical)
   */
  transformToLong(
    wideData: Array<Record<string, any>>,
    skipEmpty: boolean = true,
  ): TransformResult {
    logger.info("CSV変換開始", { wideRowCount: wideData.length });
    if (wideData.length > 0) {
      logger.debug("横持ちデータキー", { keys: Object.keys(wideData[0]).slice(0, 20) });
    }

    const longData: Array<Record<string, any>> = [];
    const errors: ValidationError[] = [];

    for (let rowIdx = 0; rowIdx < wideData.length; rowIdx++) {
      const row = wideData[rowIdx];

      // Extract common fields
      let common = this.extractCommonFields(row);

      // Validate common fields
      const [validatedCommon, commonErrors] = this.validateCommonFields(common, rowIdx);
      common = validatedCommon;
      errors.push(...commonErrors);

      // Extract details
      const details = this.extractDetails(row);

      for (let detailIdx = 0; detailIdx < details.length; detailIdx++) {
        let detail = details[detailIdx];

        // Skip empty details
        if (skipEmpty && this.isEmptyDetail(detail)) {
          continue;
        }

        // Validate detail
        const [validatedDetail, detailErrors] = this.validateDetail(detail, rowIdx, detailIdx + 1);
        detail = validatedDetail;
        errors.push(...detailErrors);

        // Merge common and detail
        const longRow = {
          ...common,
          明細番号: detailIdx + 1,
          ...detail,
        };
        longData.push(longRow);
      }
    }

    if (errors.length > 0) {
      logger.warn("CSV変換中にエラー発生", { errorCount: errors.length });
    }
    logger.info("CSV変換完了", {
      wideRowCount: wideData.length,
      longRowCount: longData.length,
      errorCount: errors.length,
    });

    return {
      long_data: longData,
      errors,
    };
  }

  private extractCommonFields(row: Record<string, any>): Record<string, any> {
    const common: Record<string, any> = {};
    for (const fieldName of COMMON_FIELDS) {
      if (fieldName in row) {
        common[fieldName] = this.normalizeValue(row[fieldName]);
      }
    }
    return common;
  }

  private extractDetails(row: Record<string, any>): Array<Record<string, any>> {
    const details: Array<Record<string, any>> = [];

    logger.debug("明細抽出開始", { columnCount: Object.keys(row).length });

    for (let n = 1; n <= this.maxDetails; n++) {
      const detail = this.extractSingleDetail(row, n);

      // Only add if detail has data
      if (!this.isEmptyDetail(detail)) {
        details.push(detail);

        // If matched without numbering for n=1, skip further details (it's a vertical row)
        if (this.isVerticalFormat(row)) {
          break;
        }
      }
    }

    logger.debug("明細抽出完了", { detailCount: details.length });
    return details;
  }

  private extractSingleDetail(row: Record<string, any>, n: number): Record<string, any> {
    const detail: Record<string, any> = {};

    // Normalize row keys (全角数字→半角数字) to match patterns
    const normalizedRow = this.createNormalizedRow(row);

    // Extract regular detail fields
    this.extractDetailFields(detail, normalizedRow, n);

    // Extract sub-detail fields (Lot No, 梱包数)
    this.extractSubDetailFields(detail, normalizedRow, n);

    return detail;
  }

  private createNormalizedRow(row: Record<string, any>): Record<string, any> {
    const normalizedRow: Record<string, any> = {};
    for (const [key, value] of Object.entries(row)) {
      const normalizedKey = this.normalizeKey(key);
      normalizedRow[normalizedKey] = value;
    }
    return normalizedRow;
  }

  private extractDetailFields(
    detail: Record<string, any>,
    normalizedRow: Record<string, any>,
    n: number,
  ): void {
    for (const fieldName of DETAIL_FIELDS) {
      const keysToTry = [`${fieldName}${n}`, `${fieldName} ${n}`];
      if (n === 1) {
        keysToTry.push(fieldName);
      }

      let matched = false;
      for (const key of keysToTry) {
        if (key in normalizedRow) {
          detail[fieldName] = this.normalizeValue(normalizedRow[key]);
          matched = true;
          break;
        }
      }

      if (!matched && n === 1) {
        // Show what similar keys exist in normalized row (debug only)
        const similarKeys = Object.keys(normalizedRow).filter((k) =>
          k.toLowerCase().includes(fieldName.toLowerCase()),
        );
        if (similarKeys.length > 0) {
          logger.debug(`フィールド "${fieldName}" 未検出`, { triedKeys: keysToTry, similarKeys });
        }
      }
    }
  }

  private extractSubDetailFields(
    detail: Record<string, any>,
    normalizedRow: Record<string, any>,
    n: number,
  ): void {
    for (const subField of SUB_DETAIL_FIELDS) {
      for (let subN = 1; subN <= 4; subN++) {
        const keysToTry = [`${subField}${n}-${subN}`, `${subField} ${n}-${subN}`];
        if (n === 1) {
          keysToTry.push(`${subField}-${subN}`);
        }

        for (const key of keysToTry) {
          if (key in normalizedRow) {
            detail[`${subField}${subN}`] = this.normalizeValue(normalizedRow[key]);
            break;
          }
        }
      }
    }
  }

  private normalizeKey(key: string): string {
    // Normalize column names: 全角数字→半角数字, trim whitespace
    const trimmed = key.trim();

    // Full-width to half-width number conversion
    const fullWidthNumbers = "０１２３４５６７８９";
    const halfWidthNumbers = "0123456789";

    let result = "";
    for (const char of trimmed) {
      const idx = fullWidthNumbers.indexOf(char);
      result += idx >= 0 ? halfWidthNumbers[idx] : char;
    }

    return result;
  }

  private isVerticalFormat(row: Record<string, any>): boolean {
    return DETAIL_FIELDS.some((f) => !(f + "1" in row) && !(f + " 1" in row) && f in row);
  }

  private isEmptyDetail(detail: Record<string, any>): boolean {
    // Check if ALL detail fields are empty
    // If ANY field has a value, it is NOT empty

    const detailKeys = Object.keys(detail);

    for (const key of detailKeys) {
      if (key.startsWith("エラー_")) continue; // Ignore error flags
      const value = detail[key];
      const trimmedValue = value ? String(value).trim() : "";

      if (trimmedValue !== "") {
        return false;
      }
    }

    return true;
  }

  private normalizeValue(value: any): string {
    if (value === null || value === undefined) {
      return "";
    }

    const s = String(value).trim();

    // Full-width to half-width conversion (numbers, letters, some symbols)
    const fullWidth =
      "０１２３４５６７８９ＡＢＣＤＥＦＧＨＩＪＫＬＭＮＯＰＱＲＳＴＵＶＷＸＹＺａｂｃｄｅｆｇｈｉｊｋｌｍｎｏｐｑｒｓｔｕｖｗｘｙｚ";
    const halfWidth = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

    let result = "";
    for (const char of s) {
      const idx = fullWidth.indexOf(char);
      result += idx >= 0 ? halfWidth[idx] : char;
    }

    return result;
  }

  private validateCommonFields(
    common: Record<string, any>,
    rowIdx: number,
  ): [Record<string, any>, ValidationError[]] {
    const errors: ValidationError[] = [];

    // Validate 発行日 (issue date)
    if ("発行日" in common) {
      const original = common["発行日"];
      const [parsed, hasError] = this.parseDate(original);
      if (hasError) {
        errors.push({
          row: rowIdx,
          field: "発行日",
          message: "日付形式が不正です",
          value: original || null,
        });
        common["エラー_発行日形式"] = 1;
      } else {
        common["発行日"] = parsed;
        common["エラー_発行日形式"] = 0;
      }
    }

    // Validate 納入日 (delivery date)
    if ("納入日" in common) {
      const original = common["納入日"];
      const [parsed, hasError] = this.parseDate(original);
      if (hasError) {
        errors.push({
          row: rowIdx,
          field: "納入日",
          message: "日付形式が不正です",
          value: original || null,
        });
        common["エラー_納入日形式"] = 1;
      } else {
        common["納入日"] = parsed;
        common["エラー_納入日形式"] = 0;
      }
    }

    return [common, errors];
  }

  private validateDetail(
    detail: Record<string, any>,
    rowIdx: number,
    detailIdx: number,
  ): [Record<string, any>, ValidationError[]] {
    const errors: ValidationError[] = [];

    // Validate 次区 (first character must be alphabet)
    if ("次区" in detail) {
      const jiku = detail["次区"];
      if (jiku && !/^[A-Za-z]/.test(jiku)) {
        errors.push({
          row: rowIdx,
          field: `次区(明細${detailIdx})`,
          message: "次区は先頭がアルファベットである必要があります",
          value: jiku || null,
        });
        detail["エラー_次区形式"] = 1;
      } else {
        detail["エラー_次区形式"] = 0;
      }
    }

    // Validate 納入量 (quantity must be non-negative number)
    if ("納入量" in detail) {
      const quantity = detail["納入量"];
      const [parsed, hasError] = this.parseQuantity(quantity);
      if (hasError) {
        errors.push({
          row: rowIdx,
          field: `納入量(明細${detailIdx})`,
          message: "納入量は0以上の数値である必要があります",
          value: quantity || null,
        });
        detail["エラー_納入量"] = 1;
      } else {
        detail["納入量"] = parsed;
        detail["エラー_納入量"] = 0;
      }
    }

    return [detail, errors];
  }

  private parseDate(value: string): [string, boolean] {
    if (!value) {
      return ["", false];
    }

    const s = this.normalizeValue(value);

    // Try various date formats
    const formats = [
      { regex: /^(\d{4})\/(\d{1,2})\/(\d{1,2})$/, parts: 3 },
      { regex: /^(\d{4})-(\d{1,2})-(\d{1,2})$/, parts: 3 },
      { regex: /^(\d{4})年(\d{1,2})月(\d{1,2})日$/, parts: 3 },
      { regex: /^(\d{4})\.(\d{1,2})\.(\d{1,2})$/, parts: 3 },
      { regex: /^(\d{4})\/(\d{1,2})$/, parts: 2 },
      { regex: /^(\d{4})-(\d{1,2})$/, parts: 2 },
      { regex: /^(\d{4})年(\d{1,2})月$/, parts: 2 },
      { regex: /^(\d{4})年$/, parts: 1 },
      { regex: /^(\d{4})$/, parts: 1 },
    ];

    for (const { regex, parts } of formats) {
      const match = s.match(regex);
      if (match) {
        return this.formatMatchedDate(match, parts);
      }
    }

    return ["", true];
  }

  private formatMatchedDate(match: RegExpMatchArray, parts: number): [string, boolean] {
    const year = parseInt(match[1], 10);
    const month = parts >= 2 ? parseInt(match[2], 10) : 1;
    const day = parts >= 3 ? parseInt(match[3], 10) : 1;

    // Range check
    if (
      year < 1900 ||
      year > 9999 ||
      (parts >= 2 && (month < 1 || month > 12)) ||
      (parts >= 3 && (day < 1 || day > 31))
    ) {
      return ["", true];
    }

    // Format as YYYY/MM/DD
    const formatted = `${year}/${String(month).padStart(2, "0")}/${String(day).padStart(2, "0")}`;
    return [formatted, false];
  }

  private parseQuantity(value: string): [string, boolean] {
    if (!value) {
      return ["", false];
    }

    let s = this.normalizeValue(value);

    // Remove commas (1,000 → 1000)
    s = s.replace(/,/g, "");

    // Validate numeric format (integer or decimal)
    // Supports: 123, 123.45, .45, 123.
    if (!/^-?\d*\.?\d*$/.test(s) || s === "." || s === "-" || s === "-.") {
      return [s, true];
    }

    const num = parseFloat(s);
    if (isNaN(num) || num < 0) {
      return [s, true];
    }

    // Return as integer if whole number
    if (Number.isInteger(num)) {
      return [String(Math.floor(num)), false];
    }

    return [String(num), false];
  }
}
